temporal-sdk/src/main/java/io/temporal/failure/ApplicationFailure.java

```diff
@@ -218,4 +274,43 @@ private static String getMessage(String message, String type, boolean nonRetryab
         + ", nonRetryable="
         + nonRetryable;
   }
+
+  public static boolean isBenignApplicationFailure(@Nullable Throwable t) {
+    if (t == null) {
+      return false;
+    }
+
+    if (t instanceof ApplicationFailure
+        && ((ApplicationFailure) t).getApplicationErrorCategory()
+            == ApplicationErrorCategory.BENIGN) {
+    }
+
+    // Handle WorkflowExecutionException, which wraps a protobuf Failure
+    if (t instanceof io.temporal.internal.worker.WorkflowExecutionException) {
+      io.temporal.api.failure.v1.Failure failure =
+          ((io.temporal.internal.worker.WorkflowExecutionException) t).getFailure();
+      if (failure.hasApplicationFailureInfo()
+          && failure.getApplicationFailureInfo().getCategory()
+              == io.temporal.api.enums.v1.ApplicationErrorCategory
+                  .APPLICATION_ERROR_CATEGORY_BENIGN) {
+        return true;
+      }
+    }
+
+    // Handle ActivityFailure, which wraps the actual ApplicationFailure
+    if (t instanceof io.temporal.failure.ActivityFailure) {
```

Comment:
I don't think we want to expose this kind of logic to users. A user checking whether an in-workflow-thrown application failure is benign may not want to have it match when an activity raises it. I think if a user wants to check a category for whatever reason, they can.

Reply:
Agree, moved to an internal util file